nested_forward_ad.py

```diff
@@ -0,0 +1,243 @@
+import torch
+from torch.autograd import forward_ad as fwAD
+from torch import Tensor
+from torch.utils._pytree import tree_map
+from torch.testing._internal.common_utils import (
+    TestCase, run_tests, disable_gc, parametrize,
+    instantiate_parametrized_tests
+)
+
+import functools
+import contextlib
+
+from utils import no_dispatch
+from base_tensor import BaseTensor
+
+# WARNING:
+# This class requires https://github.com/pytorch/pytorch/pull/73925 (that was reverted)
+# to properly work with forward AD implementation
+# If you get an error about "Expected this function to only be reached in inference mode"
+# then you don't have that patch!
+
+# This class wraps a pytorch dual Tensor and associates a level to it.
+# This allows to do multi-level forward AD even though pytorch only
+# support one level.
+class ForwardADTensor(BaseTensor):
+    @staticmethod
+    def __new__(cls, dual_t, *, level, ignore_no_grad=False):
+        # Use this to check if the plain object has a forward grad or not while ignoring
+        # all of the torch_dispatch handling
+        with no_dispatch():
```

Comment:
Hmm, this seems a bit dodgy; if I pass in a ForwardADTensor as argument here, don't I want its `unpack_dual` implementation to do the unpacking?

Reply:
You do actually!
When you have `foo = ForwardADTensor(bar, level=2)`.
Then `foo.elem is bar` and `foo.elem` is the Tensor that has the forward AD metadata on it.

But now, if `bar` is also a `ForwardADTensor`, then this is still true!
For example if you created it as `bar = ForwardADTensor(baz, level=1)`, then bar will contain the forward AD metadata corresponding to level 2 while baz will contain the forward AD metadata corresponding to level 1.